File: lib/services/notification_service.dart
import 'package:flutter/foundation.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:permission_handler/permission_handler.dart';
import 'package:timezone/timezone.dart' as tz;
import 'package:timezone/data/latest.dart' as tz;
import '../constants/app_constants.dart';
import '../models/anime.dart';

class NotificationService {
  static final NotificationService _instance = NotificationService._internal();
  factory NotificationService() => _instance;
  NotificationService._internal();

  final FlutterLocalNotificationsPlugin _localNotifications = 
      FlutterLocalNotificationsPlugin();
  final FirebaseMessaging _firebaseMessaging = FirebaseMessaging.instance;

  bool _initialized = false;
  String? _fcmToken;

  // Getters
  bool get isInitialized => _initialized;
  String? get fcmToken => _fcmToken;

  // Initialize notification service
  Future<void> initialize() async {
    if (_initialized) return;

    try {
      // Initialize timezone
      tz.initializeTimeZones();

      // Initialize local notifications
      await _initializeLocalNotifications();

      // Initialize Firebase messaging
      await _initializeFirebaseMessaging();

      _initialized = true;
    } catch (e) {
      print('Error initializing notification service: $e');
    }
  }

  // Initialize local notifications
  Future<void> _initializeLocalNotifications() async {
    const androidSettings = AndroidInitializationSettings('@mipmap/ic_launcher');
    const iosSettings = DarwinInitializationSettings(
      requestAlertPermission: true,
      requestBadgePermission: true,
      requestSoundPermission: true,
    );

    const initSettings = InitializationSettings(
      android: androidSettings,
      iOS: iosSettings,
    );

    await _localNotifications.initialize(
      initSettings,
      onDidReceiveNotificationResponse: _onNotificationTapped,
    );

    // Create notification channels for Android
    if (defaultTargetPlatform == TargetPlatform.android) {
      await _createNotificationChannels();
    }
  }

  // Create notification channels for Android
  Future<void> _createNotificationChannels() async {
    const episodeChannel = AndroidNotificationChannel(
      AppConstants.episodeNotificationChannelId,
      AppConstants.episodeNotificationChannelName,
      description: AppConstants.episodeNotificationChannelDescription,
      importance: Importance.high,
      sound: RawResourceAndroidNotificationSound('notification'),
      enableVibration: true,
      vibrationPattern: Int64List.fromList([0, 500, 250, 500]),
    );

    const generalChannel = AndroidNotificationChannel(
      AppConstants.generalNotificationChannelId,
      AppConstants.generalNotificationChannelName,
      description: AppConstants.generalNotificationChannelDescription,
      importance: Importance.defaultImportance,
    );

    await _localNotifications
        .resolvePlatformSpecificImplementation<AndroidFlutterLocalNotificationsPlugin>()
        ?.createNotificationChannel(episodeChannel);

    await _localNotifications
        .resolvePlatformSpecificImplementation<AndroidFlutterLocalNotificationsPlugin>()
        ?.createNotificationChannel(generalChannel);
  }

  // Initialize Firebase messaging
  Future<void> _initializeFirebaseMessaging() async {
    // Request permission
    final notificationSettings = await _firebaseMessaging.requestPermission(
      alert: true,
      badge: true,
      sound: true,
      carPlay: false,
      criticalAlert: false,
      provisional: false,
      announcement: false,
    );

    if (notificationSettings.authorizationStatus == AuthorizationStatus.authorized ||
        notificationSettings.authorizationStatus == AuthorizationStatus.provisional) {
      print('User granted permission');
    } else {
      print('User declined or has not accepted permission');
    }

    // Get FCM token
    _fcmToken = await _firebaseMessaging.getToken();
    print('FCM Token: $_fcmToken');

    // Listen for token refresh
    _firebaseMessaging.onTokenRefresh.listen((token) {
      _fcmToken = token;
      print('FCM Token refreshed: $token');
      // TODO: Update token in backend
    });

    // Handle foreground messages
    FirebaseMessaging.onMessage.listen(_handleForegroundMessage);

    // Handle background message taps
    FirebaseMessaging.onMessageOpenedApp.listen(_handleMessageOpenedApp);

    // Handle initial message (when app is opened from notification)
    final initialMessage = await _firebaseMessaging.getInitialMessage();
    if (initialMessage != null) {
      _handleMessageOpenedApp(initialMessage);
    }
  }

  // Handle foreground messages
  Future<void> _handleForegroundMessage(RemoteMessage message) async {
    print('Received foreground message: ${message.notification?.title}');
    
    // Show local notification for foreground messages
    await showNotification(
      title: message.notification?.title ?? 'New Episode',
      body: message.notification?.body ?? 'A new episode is available!',
      payload: message.data.toString(),
    );
  }

  // Handle message opened app
  void _handleMessageOpenedApp(RemoteMessage message) {
    print('Message opened app: ${message.notification?.title}');
    // TODO: Navigate to appropriate screen based on message data
  }

  // Handle notification tap
  void _onNotificationTapped(NotificationResponse response) {
    print('Notification tapped: ${response.payload}');
    // TODO: Handle notification tap navigation
  }

  // Request notification permissions
  Future<bool> requestPermissions() async {
    if (defaultTargetPlatform == TargetPlatform.android) {
      final status = await Permission.notification.request();
      return status.isGranted;
    } else if (defaultTargetPlatform == TargetPlatform.iOS) {
      final settings = await _firebaseMessaging.requestPermission();
      return settings.authorizationStatus == AuthorizationStatus.authorized;
    }
    return false;
  }

  // Check if notifications are enabled
  Future<bool> areNotificationsEnabled() async {
    if (defaultTargetPlatform == TargetPlatform.android) {
      return await Permission.notification.isGranted;
    } else if (defaultTargetPlatform == TargetPlatform.iOS) {
      final settings = await _firebaseMessaging.getNotificationSettings();
      return settings.authorizationStatus == AuthorizationStatus.authorized;
    }
    return false;
  }

  // Show immediate notification
  Future<void> showNotification({
    required String title,
    required String body,
    String? payload,
    int? id,
    String? imageUrl,
  }) async {
    final notificationId = id ?? DateTime.now().millisecondsSinceEpoch ~/ 1000;

    const androidDetails = AndroidNotificationDetails(
      AppConstants.episodeNotificationChannelId,
      AppConstants.episodeNotificationChannelName,
      channelDescription: AppConstants.episodeNotificationChannelDescription,
      importance: Importance.high,
      priority: Priority.high,
      showWhen: true,
      styleInformation: BigTextStyleInformation(''),
    );

    const iosDetails = DarwinNotificationDetails(
      presentAlert: true,
      presentBadge: true,
      presentSound: true,
    );

    const details = NotificationDetails(
      android: androidDetails,
      iOS: iosDetails,
    );

    await _localNotifications.show(
      notificationId,
      title,
      body,
      details,
      payload: payload,
    );
  }

  // Schedule episode notification
  Future<void> scheduleEpisodeNotification({
    required Anime anime,
    required int episode,
    required DateTime airTime,
    int offsetMinutes = 0,
  }) async {
    final notificationTime = airTime.subtract(Duration(minutes: offsetMinutes));
    
    // Don't schedule if time is in the past
    if (notificationTime.isBefore(DateTime.now())) {
      return;
    }

    final notificationId = anime.id * 1000 + episode;
    final title = 'New Episode Available!';
    final body = '${anime.displayTitle} Episode $episode is now airing!';

    const androidDetails = AndroidNotificationDetails(
      AppConstants.episodeNotificationChannelId,
      AppConstants.episodeNotificationChannelName,
      channelDescription: AppConstants.episodeNotificationChannelDescription,
      importance: Importance.high,
      priority: Priority.high,
      showWhen: true,
      styleInformation: BigTextStyleInformation(''),
      icon: '@mipmap/ic_launcher',
    );

    const iosDetails = DarwinNotificationDetails(
      presentAlert: true,
      presentBadge: true,
      presentSound: true,
    );

    const details = NotificationDetails(
      android: androidDetails,
      iOS: iosDetails,
    );

    await _localNotifications.zonedSchedule(
      notificationId,
      title,
      body,
      tz.TZDateTime.from(notificationTime, tz.local),
      details,
      payload: 'episode_${anime.id}_$episode',
      uiLocalNotificationDateInterpretation:
          UILocalNotificationDateInterpretation.absoluteTime,
      matchDateTimeComponents: DateTimeComponents.time,
    );

    print('Scheduled notification for ${anime.displayTitle} Episode $episode at $notificationTime');
  }

  // Cancel scheduled notification
  Future<void> cancelEpisodeNotification({
    required int animeId,
    required int episode,
  }) async {
    final notificationId = animeId * 1000 + episode;
    await _localNotifications.cancel(notificationId);
    print('Cancelled notification for anime $animeId episode $episode');
  }

  // Cancel all notifications for an anime
  Future<void> cancelAnimeNotifications(int animeId) async {
    final pendingNotifications = await _localNotifications.pendingNotificationRequests();
    
    for (final notification in pendingNotifications) {
      if (notification.id >= animeId * 1000 && notification.id < (animeId + 1) * 1000) {
        await _localNotifications.cancel(notification.id);
      }
    }
    
    print('Cancelled all notifications for anime $animeId');
  }

  // Get pending notifications
  Future<List<PendingNotificationRequest>> getPendingNotifications() async {
    return await _localNotifications.pendingNotificationRequests();
  }

  // Cancel all notifications
  Future<void> cancelAllNotifications() async {
    await _localNotifications.cancelAll();
    print('Cancelled all notifications');
  }

  // Subscribe to topic (for general anime news)
  Future<void> subscribeToTopic(String topic) async {
    try {
      await _firebaseMessaging.subscribeToTopic(topic);
      print('Subscribed to topic: $topic');
    } catch (e) {
      print('Error subscribing to topic $topic: $e');
    }
  }

  // Unsubscribe from topic
  Future<void> unsubscribeFromTopic(String topic) async {
    try {
      await _firebaseMessaging.unsubscribeFromTopic(topic);
      print('Unsubscribed from topic: $topic');
    } catch (e) {
      print('Error unsubscribing from topic $topic: $e');
    }
  }

  // Show episode reminder notification
  Future<void> showEpisodeReminder({
    required Anime anime,
    required int episode,
    required Duration timeUntil,
  }) async {
    final title = 'Episode Reminder';
    final body = '${anime.displayTitle} Episode $episode airs in ${_formatDuration(timeUntil)}!';
    
    await showNotification(
      title: title,
      body: body,
      payload: 'reminder_${anime.id}_$episode',
    );
  }

  // Format duration for display
  String _formatDuration(Duration duration) {
    if (duration.inDays > 0) {
      return '${duration.inDays} day${duration.inDays > 1 ? 's' : ''}';
    } else if (duration.inHours > 0) {
      return '${duration.inHours} hour${duration.inHours > 1 ? 's' : ''}';
    } else if (duration.inMinutes > 0) {
      return '${duration.inMinutes} minute${duration.inMinutes > 1 ? 's' : ''}';
    } else {
      return 'less than a minute';
    }
  }

  // Dispose resources
  void dispose() {
    // Clean up any resources if needed
  }
}

File: src/services/notificationService.ts
import PushNotification from 'react-native-push-notification';
import PushNotificationIOS from '@react-native-community/push-notification-ios';
import { Platform } from 'react-native';
import { StorageService } from './storageService';
import { Anime, Episode } from '../types/index';

export class NotificationService {
  private static isInitialized = false;

  static initialize(): void {
    if (this.isInitialized) return;

    PushNotification.configure({
      onRegister: function (token) {
        console.log('TOKEN:', token);
      },

      onNotification: function (notification) {
        console.log('NOTIFICATION:', notification);

        if (Platform.OS === 'ios') {
          notification.finish(PushNotificationIOS.FetchResult.NoData);
        }
      },

      onAction: function (notification) {
        console.log('ACTION:', notification.action);
        console.log('NOTIFICATION:', notification);
      },

      onRegistrationError: function (err) {
        console.error(err.message, err);
      },

      permissions: {
        alert: true,
        badge: true,
        sound: true,
      },

      popInitialNotification: true,
      requestPermissions: Platform.OS === 'ios',
    });

    if (Platform.OS === 'android') {
      PushNotification.createChannel(
        {
          channelId: 'anime-episodes',
          channelName: 'Anime Episodes',
          channelDescription: 'Notifications for new anime episodes',
          playSound: true,
          soundName: 'default',
          importance: 4,
          vibrate: true,
        },
        (created) => console.log(`Channel created: ${created}`)
      );
    }

    this.isInitialized = true;
  }

  static async requestPermissions(): Promise<boolean> {
    return new Promise((resolve) => {
      if (Platform.OS === 'ios') {
        PushNotificationIOS.requestPermissions({
          alert: true,
          badge: true,
          sound: true,
        }).then((permissions) => {
          resolve(permissions.alert || permissions.badge || permissions.sound);
        });
      } else {
        PushNotification.requestPermissions().then((permissions) => {
          resolve(!!permissions);
        });
      }
    });
  }

  static async scheduleEpisodeNotification(
    anime: Anime,
    episode: Episode,
    delay: number = 0
  ): Promise<void> {
    try {
      const settings = await StorageService.getAppSettings();
      
      if (!settings.notifications.enabled) {
        return;
      }

      const notificationTime = new Date(episode.airingAt * 1000);
      
      // Apply delay based on user settings
      switch (settings.notifications.notificationTime) {
        case '1hour':
          notificationTime.setHours(notificationTime.getHours() + 1);
          break;
        case '24hours':
          notificationTime.setDate(notificationTime.getDate() + 1);
          break;
        // 'immediate' needs no delay
      }

      const now = new Date();
      if (notificationTime <= now) {
        // Episode already aired, send immediate notification
        this.sendImmediateNotification(anime, episode);
        return;
      }

      const title = `New Episode Available!`;
      const message = `${anime.title.english || anime.title.romaji} - Episode ${episode.number}`;

      const notificationId = `${anime.id}-${episode.number}`;

      PushNotification.localNotificationSchedule({
        id: notificationId,
        title,
        message,
        date: notificationTime,
        playSound: settings.notifications.soundEnabled,
        vibrate: settings.notifications.vibrationEnabled,
        channelId: 'anime-episodes',
        actions: ['View', 'Dismiss'],
        userInfo: {
          animeId: anime.id,
          episodeNumber: episode.number,
          type: 'episode_release',
        },
        ...(Platform.OS === 'android' && {
          largeIcon: 'ic_launcher',
          smallIcon: 'ic_notification',
        }),
      });

      console.log(`Scheduled notification for ${anime.title.romaji} Episode ${episode.number}`);
    } catch (error) {
      console.error('Error scheduling notification:', error);
    }
  }

  static sendImmediateNotification(anime: Anime, episode: Episode): void {
    try {
      const title = `New Episode Available!`;
      const message = `${anime.title.english || anime.title.romaji} - Episode ${episode.number}`;

      PushNotification.localNotification({
        title,
        message,
        playSound: true,
        vibrate: true,
        channelId: 'anime-episodes',
        actions: ['View', 'Dismiss'],
        userInfo: {
          animeId: anime.id,
          episodeNumber: episode.number,
          type: 'episode_release',
        },
        ...(Platform.OS === 'android' && {
          largeIcon: 'ic_launcher',
          smallIcon: 'ic_notification',
        }),
      });
    } catch (error) {
      console.error('Error sending immediate notification:', error);
    }
  }

  static cancelAnimeNotifications(animeId: number): void {
    try {
      PushNotification.getScheduledLocalNotifications((notifications) => {
        notifications.forEach((notification) => {
          if (notification.userInfo?.animeId === animeId) {
            PushNotification.cancelLocalNotification(notification.id);
          }
        });
      });
    } catch (error) {
      console.error('Error canceling notifications:', error);
    }
  }

  static cancelEpisodeNotification(animeId: number, episodeNumber: number): void {
    try {
      const notificationId = `${animeId}-${episodeNumber}`;
      PushNotification.cancelLocalNotification(notificationId);
    } catch (error) {
      console.error('Error canceling episode notification:', error);
    }
  }

  static clearAllNotifications(): void {
    try {
      PushNotification.cancelAllLocalNotifications();
    } catch (error) {
      console.error('Error clearing all notifications:', error);
    }
  }

  static async getScheduledNotifications(): Promise<any[]> {
    return new Promise((resolve) => {
      PushNotification.getScheduledLocalNotifications((notifications) => {
        resolve(notifications);
      });
    });
  }

  static async rescheduleAllNotifications(): Promise<void> {
    try {
      // Clear existing notifications first
      this.clearAllNotifications();

      // Get user's anime list
      const userAnimeList = await StorageService.getUserAnimeList();
      
      // Reschedule notifications for each anime with upcoming episodes
      for (const userAnime of userAnimeList) {
        if (userAnime.notificationsEnabled && userAnime.anime.nextAiringEpisode) {
          const episode: Episode = {
            number: userAnime.anime.nextAiringEpisode.episode,
            airingAt: userAnime.anime.nextAiringEpisode.airingAt,
            animeId: userAnime.anime.id,
          };

          await this.scheduleEpisodeNotification(userAnime.anime, episode);
        }
      }
    } catch (error) {
      console.error('Error rescheduling notifications:', error);
    }
  }

  static async testNotification(): Promise<void> {
    try {
      PushNotification.localNotification({
        title: 'Test Notification',
        message: 'This is a test notification for Anime Episode Alert',
        playSound: true,
        vibrate: true,
        channelId: 'anime-episodes',
      });
    } catch (error) {
      console.error('Error sending test notification:', error);
    }
  }
}